geo_tag/geo_tag.py

```diff
@@ -0,0 +1,221 @@
+        with open('us_state_abbrev.json') as json_file:
+            self.abbrev_us_state = json.load(json_file)
+        # self.city_data is a big json from cloudberry -> cache
+        city_datafile = open("city.json", "r", encoding='UTF-8')
+        self.city_data = json.load(city_datafile)
+        city_datafile.close()
+
+    def init_geo_cache(self):
+        # check pickle module
+        filename = "cache_geo_hashmap"
+        if os.path.exists(filename):
+            cache_file = open(filename, 'rb')
+            self.cache = pickle.load(cache_file)
+            return
+        else:
+            city_json = self.city_data
+            feature = city_json.get('features')
+            for f in feature:
+                item = f.get('properties')
+                # store the city, state as key; (cityID, countyID, stateID, countyName) as value.
+                if item is not None and "name" in item and "stateName" in item:
+                    _city_state = item["name"] + ", " + item["stateName"]
+                    # key, value
+                    self.cache[_city_state] = (item["cityID"], item["countyID"], item["stateID"], item["countyName"])
+
+                else:
+                    raise KeyError("can not build cache from city.json.")
+
+            # store it using pickle
+            cache_file = open(filename, 'wb')
+            pickle.dump(self.cache, cache_file)
+            cache_file.close()
+
+    def get_coordinate(self, tweet_json: dict, mode: int):
+        coord = tweet_json.get('coordinates')
+        place = tweet_json.get('place')
+
+        if coord is not None:
+            return coord.get('coordinates')
+        else:
+            # 2. get the central point of the bounding_box
+            if place is not None:
+                bounding_box = place.get('bounding_box').get('coordinates')[0]
+                ''' Twitter has some wield format historically, though it still rectangle, 
+                but it is not always in (sw, se, ne, nw) order
+                '''
+                swLog = min(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                swLat = min(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
+                neLog = max(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                neLat = max(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
+
+                # AsterixDB is unhappy with this kind of point "rectangular"
+                if (swLog == neLog and swLat == neLat):
+                    swLog = neLog - 0.0000001
+                    swLat = neLat - 0.0000001
+
+                if (swLog > neLog or swLat > neLat):
+                    raise ValueError("Invalid coordinates in bounding_box.")
+
+                # Return Central Point or Random Point in the polygon each time.
+                if mode == 0:
```

Comment:
define mode 0 and 1 for reference. give a variable name. or use a boolean value `randomize` instead an int value `mode`

Reply:
wait, I thought we need to have several different randomize configuration,

- No random (exact)
- uniform distribution random
- normal distribution random (most important one)